File: cpp/lib/telegraph/local/dummy_device.cpp
#include "dummy_device.hpp"

#include <boost/asio/io_context.hpp>
#include <boost/asio/deadline_timer.hpp>
#include <boost/asio/spawn.hpp>

namespace telegraph {
    dummy_device::dummy_device(io::io_context& ioc, const std::string_view& name, 
                                std::unique_ptr<node>&& tree)
                           : local_context(ioc, name, "dummy_device", 
                                    params{}, std::move(tree))  {}

    dummy_device::~dummy_device() {}
    void
    dummy_device::add_publisher(const variable* v, const publisher_ptr& p) {
        publishers_.emplace(v, p);
    }

    void
    dummy_device::add_handler(const action* a, const dummy_device::handler& h) {
        handlers_.emplace(a, h);
    }

    params_stream_ptr 
    dummy_device::request(io::yield_ctx&, const params& p) {
        return nullptr;
    }

    subscription_ptr
    dummy_device::subscribe(io::yield_ctx&, const variable* v,
                            float min_interval, float max_interval,
                            float timeout) {
        auto it = publishers_.find(v);
        if (it == publishers_.end()) return nullptr;
        auto p = it->second;
        if (!p) return nullptr;
        return p->subscribe(min_interval, max_interval);
    }

    value
    dummy_device::call(io::yield_ctx& yield, action* a,
                            value arg, float timeout) {
        auto it = handlers_.find(a);
        if ( it == handlers_.end()) return value::invalid();
        return it->second(yield, arg);
    }

   local_context_ptr 
    dummy_device::create(io::yield_ctx&, io::io_context& ioc,
            const std::string_view& name, const std::string_view& type,
            const params& p) {
        auto childA = new variable(2, "a", "A", "", value_type::Float);
        auto childB = new variable(3, "b", "B", "", value_type::Uint8);

        std::vector<std::string> labels{"On", "Off"};
        auto status_type = value_type("Status", std::move(labels));
        auto childC = new variable(4, "c", "C", "", status_type);
        
        auto childD = new variable(5, "d", "D", "", value_type::Bool);


        std::vector<std::string> test_labels = {"option1", "option2", "option3"};
        value_type test_enum = value_type("test_enum", test_labels);
        std::vector<std::string> short_labels = {"first", "second"};
        value_type short_enum = value_type("short_enum", short_labels);
        // action
        auto action1 = new action(1, "action1", "action1", "", value_type::Uint8, value_type::Uint8);
        auto action2 = new action(2, "action2", "action2", "", value_type::Bool, value_type::Bool);
        auto action3 = new action(3, "action3", "action3", "", value_type::None, value_type::Uint8);
        auto action4 = new action(4, "action4", "action4", "", test_enum, test_enum);
        auto action5 = new action(5, "action5", "action5", "", short_enum, short_enum);

        std::vector<node*> children{childA, childB, childC, childD, action1, action2, action3, action4, action5};
        auto root = std::make_unique<group>(1, "foo", "Foo", "", "", 1, std::move(children));
        auto dev = std::make_shared<dummy_device>(ioc, name, std::move(root));

        auto a_publisher = std::make_shared<publisher>(ioc, value_type::Float);
        auto b_publisher = std::make_shared<publisher>(ioc, value_type::Int32);
        auto c_publisher= std::make_shared<publisher>(ioc, status_type);
        auto d_publisher= std::make_shared<publisher>(ioc, value_type::Bool);


        dev->add_publisher(childA, a_publisher);
        dev->add_publisher(childB, b_publisher);
        dev->add_publisher(childC, c_publisher);
        dev->add_publisher(childD, d_publisher);


        // add action
        dev->add_handler(action1, [] (io::yield_ctx& yield, value val) -> value {
            return value{(uint8_t)(unwrap<uint8_t>(val) * 2)};
        });
        dev->add_handler(action2, [] (io::yield_ctx& yield, value val) -> value {
            return value{(bool) !unwrap<bool>(val)};
        });
        dev->add_handler(action3, [] (io::yield_ctx& yield, value val) -> value {
            return value{(uint8_t) 42};
        });
        dev->add_handler(action4, [] (io::yield_ctx& yield, value val) -> value {
            uint8_t num = unwrap<uint8_t>(val);
            return value(value_type::Enum, num);
        });
        dev->add_handler(action5, [] (io::yield_ctx& yield, value val) -> value {
            uint8_t num = unwrap<uint8_t>(val);
            return value(value_type::Enum, num);
        });

        // spawn data-pushing loop....
        std::weak_ptr<dummy_device> w{dev};

        // spawn loop for a
        io::spawn(ioc, [&ioc, w, a_publisher](io::yield_context yield) {
            // create a timer
            io::deadline_timer timer{ioc};
            while (true) {
                timer.expires_from_now(boost::posix_time::seconds(2));
                timer.async_wait(yield);
                { auto dev = w.lock(); if (!dev) break; }
                (*a_publisher) << value{1.0f}; 

                timer.expires_from_now(boost::posix_time::seconds(2));
                timer.async_wait(yield);
                { auto dev = w.lock(); if (!dev) break; }
                (*a_publisher) << value{2.5f};
                timer.expires_from_now(boost::posix_time::seconds(2));
                timer.async_wait(yield);
                { auto dev = w.lock(); if (!dev) break; }
                (*a_publisher) << value{3.8f};
            }
        });
        io::spawn(ioc, [&ioc, w, b_publisher](io::yield_context yield) {
            // create a timer
            io::deadline_timer timer{ioc};
            uint8_t val= 0;
            while (true) {
                timer.expires_from_now(boost::posix_time::milliseconds(5));
                timer.async_wait(yield);
                { auto dev = w.lock(); if (!dev) break; }
                (*b_publisher) << value{val};
                val++;
            }
        });

        io::spawn(ioc, [&ioc, w, d_publisher](io::yield_context yield) {
            // create a timer
            io::deadline_timer timer{ioc};
            bool val= false;
            while (true) {
                timer.expires_from_now(boost::posix_time::seconds(2));
                timer.async_wait(yield);
                { auto dev = w.lock(); if (!dev) break; }
                (*d_publisher) << value{val};
                val = !val;
            }
        });
        return dev;
    }
}

File: backend/server/src/local/dummy_device.rs
use std::collections::HashMap;
use std::sync::{Arc, Mutex, Weak};
use std::time::Duration;

use tokio::time::sleep;

use log::{debug, error};

use super::Device;
use crate::common::subscription::Publisher;
use crate::nodes::{Node, NodeID, Variable};
use crate::types;
use crate::value::Value;

pub struct DummyDevice {
    root: Node,
    // TODO: how do I want to associate publishers to variables? Maybe I want to make groups use
    // the interior mutability pattern for this reason.
    publishers: HashMap<NodeID, Arc<Mutex<Publisher>>>,
}

impl Device for DummyDevice {
    type Handler = ();

    fn add_publisher(&mut self, v: &Variable, p: Arc<Mutex<Publisher>>) {
        self.publishers.insert(v.info().id, p);
    }

    // This is a dummy device, so (for now) we don't care about handlers
    fn add_handler(&mut self, _v: &Variable, _p: Self::Handler) {}

    // TODO: params_stream_ptr request(io::yield_ctx&, const params& p) override;

    // TODO: subscription_ptr subscribe(io::yield_ctx& ctx,
    //     const variable* v,
    //     float min_interval, float max_interval,
    //     float timeout) override;

    // TODO: value call(io::yield_ctx& yield, action* a, value v, float timeout) override;

    // TODO: subscription_ptr subscribe(io::yield_ctx& yield,
    //     const std::vector<std::string_view>& path,
    //     float min_interval, float max_interval,
    //     float timeout) override {
    //     auto v = dynamic_cast<variable*>(tree_->from_path(path));
    //     if (!v) return nullptr;
    //     return subscribe(yield, v, min_interval, max_interval, timeout);
    // }

    // TODO: value call(io::yield_ctx& yield,
    //     const std::vector<std::string_view>& path,
    //     value v, float timeout) override {
    //     auto a = dynamic_cast<action*>(tree_->from_path(path));
    //     if (!a) return value::invalid();
    //     return call(yield, a, v, timeout);
    // }

    // TODO: bool write_data(io::yield_ctx& yield,
    //     variable* v,
    //     const std::vector<datapoint>& data) override {
    //     return false;
    // }
    // TODO: bool write_data(io::yield_ctx& yield,
    //     const std::vector<std::string_view>&,
    //     const std::vector<datapoint>& data) override {
    //     return false;
    // }

    // TODO: data_query_ptr query_data(io::yield_ctx& yield,
    //     const variable* v) override {
    //     return nullptr;
    // }
    // TODO: data_query_ptr query_data(io::yield_ctx& yield,
    //     const std::vector<std::string_view>& v) override {
    //     return nullptr;
    // }

    // TODO: static local_context_ptr create(io::yield_ctx&, io::io_context& ioc,
    //     const std::string_view& name, const std::string_view& type,
    //     const params& p);
    // };
}

fn publisher_ptr(value_type: types::Type) -> Arc<Mutex<Publisher>> {
    Arc::new(Mutex::new(Publisher::new(value_type)))
}

fn write_value(publisher: Weak<Mutex<Publisher>>, value: Value) -> Option<Value> {
    let publisher = publisher.upgrade()?;

    // Setting this to a variable before returning placates the borrow checker
    let out = match publisher.lock() {
        Ok(mut p) => {
            p.update(value);
            Some(value)
        }
        Err(e) => {
            error!("error when unlocking the publisher: {}", e);
            None
        }
    };

    out
}

impl DummyDevice {
    pub fn new() -> Self {
        let child_a = Node::variable(2, "a", "A", "", types::FLOAT);
        let child_b = Node::variable(3, "b", "B", "", types::UINT8);
        let child_c = Node::variable(4, "c", "C", "", types::enum_type("Status", &["On", "Off"]));

        let mut publishers: HashMap<NodeID, Arc<Mutex<Publisher>>> = HashMap::new();
        let a_publisher = publisher_ptr(types::FLOAT);
        publishers.insert(child_a.info().id, a_publisher.clone());

        tokio::spawn(async move {
            let publisher = Arc::downgrade(&a_publisher);
            // We don't want to keep an Arc to the publisher because we'll just terminate the loop
            // if the publisher gets dropped.
            std::mem::drop(a_publisher);

            loop {
                sleep(Duration::from_secs(2)).await;

                match write_value(publisher.clone(), 1.0.into()) {
                    None => break,
                    Some(v) => debug!("Wrote {:?}", v),
                }

                sleep(Duration::from_secs(2)).await;

                match write_value(publisher.clone(), 2.5.into()) {
                    None => break,
                    Some(v) => debug!("Wrote {:?}", v),
                }

                sleep(Duration::from_secs(2)).await;

                match write_value(publisher.clone(), 3.8.into()) {
                    None => break,
                    Some(v) => debug!("Wrote {:?}", v),
                }
            }
        });

        let root = Node::group(1, "foo", "Foo", "", vec![child_a, child_b, child_c], "", 1);

        Self {
            root,
            publishers,
        }
    }
}
